### PR TITLE
Fix closing SignalR Websocket connection on server close

### LINKED ISSUE #53165: SignalR prevents .NET8 app to stop if clients are connected
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

I have a Blazor App (the repo given is from the Blazor Webassembly Standalone App with ASP.NET Core Hosted option, but this bug also occurs when using the new Blazor Web App template). In the app a signalR connection between the server and client is opened. When I close the program there are two different behaviors in .NET7 and .NET8.
- In .NET7, when i send the shut down command to the app, it shutdowns immediately and closes all opened connections.
- In .NET8, when i send the shut down command to the app, it logs that the application is shutting down but it takes ~10s to send me a message that the signalR client has disconnected and only after that it really exits the app.

I would assume that the .NET7 version is the correct behavior.

### Expected Behavior

Log in .NET7

```
12:30:51 info: SignalRProblem.NotificationHub[0]
      Notif WS connected: 8m-yY_W585vGcU7ce9IceA
12:30:54 info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
12:30:54 info: SignalRProblem.NotificationHub[0]
      Notif WS disconnected: 8m-yY_W585vGcU7ce9IceA
```

Log in .NET8
```
12:29:44 info: SignalRProblem.NotificationHub[0]
      Notif WS connected: HXvhuJ8p6LnHKe16xnU5oA
12:29:48 info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
12:30:00 info: SignalRProblem.NotificationHub[0]
      Notif WS disconnected: HXvhuJ8p6LnHKe16xnU5oA
```

Note the time difference between the "Application is shutting down" message and the "Notif WS disconnected" message

### Steps To Reproduce

[Minimal Repo](https://github.com/pedroveiga99/SignalRProblem.git)

This repo is targeting the .NET7 framework. You could just change the Server project to target the .NET8 version and the problem will occur.

### Exceptions (if any)

_No response_

### .NET Version

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/53165

Regressed in https://github.com/dotnet/aspnetcore/commit/f56c242d037e7a11626151e78333caa26c3b5d00#diff-1a6e4221ebf9343c0f831c3bddda91e203203f07d212575ad2f413ce3b3e2b34L177
We just never had a "server close" test.

### CODE REVIEW

File: src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs [hunk 1]
```diff
@@ -168,6 +168,14 @@ private async Task ExecuteAsync(HttpContext context, ConnectionDelegate connecti
             {
                 transport = HttpTransportType.WebSockets;
                 connection = await GetOrCreateConnectionAsync(context, options);
+
+                if (connection is not null)
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: how can this be null?

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: If you attempt to connect with a connection id query string param that doesn't exist on this server instance.

Comment (davidfowl) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: This method name is bad then, GetOrCreateOrNot

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: 😆 this method name is >8 years old now.

Comment (davidfowl) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: I blame myself

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs:172]: I do too 😉

**Comment by elitastic** [comment]
This is great news, thank you @BrennanConroy!

Is there a temporary workaround available to prevent a slow shutdown?

File: src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs [hunk 2]
```diff
@@ -2842,6 +2842,35 @@ public async Task WebSocketConnectionClosingTriggersConnectionClosedToken()
...
+            var context = MakeRequest("/foo", connection, services);
+            SetTransport(context, HttpTransportType.WebSockets);
+
+            var builder = new ConnectionBuilder(services.BuildServiceProvider());
+            builder.UseConnectionHandler<TestConnectionHandler>();
+            var app = builder.Build();
+            var options = new HttpConnectionDispatcherOptions();
+            options.WebSockets.CloseTimeout = TimeSpan.FromSeconds(1);
+
+            _ = dispatcher.ExecuteAsync(context, options, app);
```
Comment (davidfowl) [inline src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs:2865]: Should we be making this this task completes?

Comment (BrennanConroy) [reply src/SignalR/common/Http.Connections/test/HttpConnectionDispatcherTests.cs]: Sure

**Review by davidfowl — APPROVED** [review]

**Pull request merged**